calico/felix/fiptables.py

```diff
@@ -1,214 +1,451 @@
+
+    def truncate_rules(self, count):
+        """
+        This is a utility function to remove any excess rules from a
+        chain. After we have carefully inserted all the rules we want at the
+        start, we want to get rid of any legacy rules from the end.
+
+        It takes a chain object, and a count for how many of the rules should
+        be left in place.
+        """
+        log.debug("Truncate %s chain %s to length %d (length now %d)",
+                  self.type,
+                  self.name,
+                  count,
+                  len(self.rules))
+
+        while len(self.rules) > count:
+            self.table.ops.append([IPTABLES_CMD[self.type],
+                                   "-w",
+                                   "-t",
+                                   self.table.name,
+                                   "-D",
+                                   self.name,
+                                   str(count + 1)])
+            self.rules.pop(count)
+
+
+    def insert_rule(self, rule, position=0, force_position=True):
+        """
+        Add an iptables rule to a chain if it does not already exist. Position
+        is the position for the insert as an offset; if set to RULE_POSN_LAST
+        then the rule is appended.
+
+        If force_position is True, then the rule is added at the specified point
+        unless it already exists there. If force_position is False, then the
+        rule is added only if it does not exist anywhere in the list of rules.
+
+        Position starts at offset 0, even though iptables wants it to start at
+        offset 1.
+        """
+        if position == RULE_POSN_LAST:
+            position = len(self.rules)
+
+        if force_position:
+            if (len(self.rules) <= position) or (rule != self.rules[position]):
+                # Either adding after existing rules, or replacing an existing rule.
+                self._insert_rule(rule, position)
+        else:
+            if rule not in self.rules:
+                self._insert_rule(rule, position)
+        return
+
+    def _insert_rule(self, rule, position):
+        """
+        Insert the given rule at the supplied position.
+        
+        position of 0 means "insert at the start", which is what iptables
+        considers as position 1.
+
+        This method is private; callers outside this module should call
+        insert_rule instead.
+        """
+        args = [IPTABLES_CMD[self.type],
+                "-w",
+                "-t",
+                self.table.name,
+                "-I",
+                self.name,
+                str(position + 1)]
+        args.extend(rule.generate_fields())
+        self.table.ops.append(args)
+        self.rules.insert(position, rule)
+        log.debug("Creating rule : %s", args)
+
+    def __str__(self):
+        output = "  Chain %s\n" % self.name
+        for rule in self.rules:
+            output += "    %s\n" % rule
+
```

Comment:
It is slow, but on the other hand there are not many chains or rules. I ended up cutting dramatically the number of such string concatenations, but not all (because readability got too low). We use `join` for all the big lists, but there are a few concatenations as well.